src/Control/Monad/Eff/AVar.js

```diff
@@ -0,0 +1,311 @@
+/* globals exports, setTimeout */
+/* jshint -W097 */
+
+"use strict";
+
+var EMPTY = {};
+var NO_EFFECT = function () {};
+
+function MutableQueue () {
+  this.head = null;
+  this.last = null;
+  this.size = 0;
+}
+
+function MutableCell (queue, value) {
+  this.queue = queue;
+  this.value = value;
+  this.next  = null;
+  this.prev  = null;
+}
+
+function putLast (queue, value) {
+  var cell = new MutableCell(queue, value);
+  switch (queue.size) {
+  case 0:
+    queue.head = cell;
+    break;
+  case 1:
+    cell.prev = queue.head;
+    queue.head.next = cell;
+    queue.last = cell;
+    break;
+  default:
+    cell.prev = queue.last;
+    queue.last.next = cell;
+    queue.last = cell;
+  }
+  queue.size++;
+  return cell;
+}
+
+function takeLast (queue) {
+  var cell;
+  switch (queue.size) {
+  case 0:
+    return null;
+  case 1:
+    cell = queue.head;
+    queue.head = null;
+    break;
+  case 2:
+    cell = queue.last;
+    queue.head.next = null;
+    queue.last = null;
+    break;
+  default:
+    cell = queue.last;
+    queue.last = cell.prev;
+    queue.last.next = null;
+  }
+  cell.prev = null;
+  cell.queue = null;
+  queue.size--;
+  return cell.value;
+}
+
+function takeHead (queue) {
+  var cell;
+  switch (queue.size) {
+  case 0:
+    return null;
+  case 1:
+    cell = queue.head;
+    queue.head = null;
+    break;
+  case 2:
+    cell = queue.head;
+    queue.last.prev = null;
+    queue.head = queue.last;
+    queue.last = null;
+    break;
+  default:
+    cell = queue.head;
+    queue.head = cell.next;
+    queue.head.prev = null;
+  }
+  cell.next = null;
+  cell.queue = null;
+  queue.size--;
+  return cell.value;
+}
+
+function deleteCell (cell) {
+  if (cell.queue === null) {
+    return;
+  }
+  if (cell.queue.last === cell) {
+    takeLast(cell.queue);
+    return;
+  }
+  if (cell.queue.head === cell) {
+    takeHead(cell.queue);
+    return;
+  }
+  if (cell.prev) {
+    cell.prev.next = cell.next;
+  }
+  if (cell.next) {
+    cell.next.prev = cell.prev;
+  }
+  cell.queue.size--;
+  cell.queue = null;
+  cell.value = null;
+  cell.next  = null;
+  cell.prev  = null;
+}
+
+function AVar () {
+  this.draining = false;
+  this.error    = null;
+  this.value    = EMPTY;
+  this.takes    = new MutableQueue();
+  this.reads    = new MutableQueue();
+  this.puts     = new MutableQueue();
+}
```

Comment:
Because you don't always want a value. There's no real difference. In the current case, if I want a value, I just mutate it. In the other case I'd have to provide `EMPTY`.